src/main/java/com/epam/reportportal/auth/integration/github/GitHubUserReplicator.java

```diff
@@ -108,63 +120,62 @@ public User replicateUser(String accessToken) {
 	 * @param gitHubClient Configured github client
 	 * @return Internal User representation
 	 */
+	@Transactional
 	public User replicateUser(UserResource userResource, GitHubClient gitHubClient) {
-		String login = normalizeId(userResource.login);
-		Optional<User> userOptional = userRepository.findByLogin(login);
-		boolean isExist = userOptional.isPresent();
-		User user;
-		if (!isExist) {
-			user = new User();
-			fillUserData(user, userResource, gitHubClient);
-			fillDefaultUserData(user);
-			userRepository.save(user);
-		} else {
-			user = userOptional.get();
-			if (UserType.GITHUB.equals(user.getUserType())) {
-				dataStorage.delete(user.getAttachment());
-				fillUserData(user, userResource, gitHubClient);
+		String login = normalizeId(userResource.getLogin());
+		return userRepository.findByLogin(login).map(u -> {
+			if (UserType.GITHUB.equals(u.getUserType())) {
+				dataStorage.delete(u.getAttachment());
+				updateUser(u, userResource, gitHubClient);
 			} else {
 				//if user with such login exists, but it's not GitHub user than throw an exception
-				throw new UserSynchronizationException("User with login '" + user.getId() + "' already exists");
+				throw new UserSynchronizationException("User with login '" + u.getId() + "' already exists");
 			}
+			return u;
+		}).orElseGet(() -> userRepository.save(createUser(userResource, gitHubClient)));
+	}
+
+	private void updateUser(User user, UserResource userResource, GitHubClient gitHubClient) {
+		String email = userResource.getEmail();
+		if (Strings.isNullOrEmpty(email)) {
+			email = retrieveEmail(gitHubClient).orElseThrow(() -> new UserSynchronizationException("User 'email' has not been provided"));
```

Comment:
There is a lot of duplicate code. Can you refactor it?

Reply:
updated